internal/planner/type_join.go

```diff
@@ -452,108 +452,206 @@ func fetchPrimaryDoc(node, subNode planNode, parentProp string) (bool, error) {
 	return true, nil
 }
 
-type joinDirection struct {
-	firstNode      planNode
-	secondNode     planNode
-	secondaryField immutable.Option[string]
-	primaryField   string
-	isInverted     bool
-}
-
-func (dir *joinDirection) invert() {
-	if !dir.secondaryField.HasValue() {
-		// If the secondary field has no value it cannot be inverted
-		return
-	}
-	dir.isInverted = !dir.isInverted
-	dir.firstNode, dir.secondNode = dir.secondNode, dir.firstNode
-	dir.secondaryField, dir.primaryField = immutable.Some(dir.primaryField), dir.secondaryField.Value()
-}
-
 type invertibleTypeJoin struct {
 	docMapper
 
-	root        planNode
-	subType     planNode
-	rootName    immutable.Option[string]
-	subTypeName string
+	skipChild bool
 
-	subSelect         *mapper.Select
-	subSelectFieldDef client.FieldDefinition
+	parentSide joinSide
+	childSide  joinSide
 
-	isSecondary         bool
-	secondaryFieldIndex immutable.Option[int]
 	secondaryFetchLimit uint
 
 	// docsToYield contains documents read and ready to be yielded by this node.
-	docsToYield []core.Doc
-
-	dir joinDirection
+	docsToYield       []core.Doc
+	encounteredDocIDs []string
 }
 
 func (join *invertibleTypeJoin) replaceRoot(node planNode) {
-	join.root = node
-	if join.dir.isInverted {
-		join.dir.secondNode = node
-	} else {
-		join.dir.firstNode = node
-	}
+	join.getFirstSide().plan = node
 }
 
 func (join *invertibleTypeJoin) Init() error {
-	if err := join.subType.Init(); err != nil {
+	if err := join.childSide.plan.Init(); err != nil {
 		return err
 	}
-	return join.root.Init()
+	return join.parentSide.plan.Init()
 }
 
 func (join *invertibleTypeJoin) Start() error {
-	if err := join.subType.Start(); err != nil {
+	if err := join.childSide.plan.Start(); err != nil {
 		return err
 	}
-	return join.root.Start()
+	return join.parentSide.plan.Start()
 }
 
 func (join *invertibleTypeJoin) Close() error {
-	if err := join.root.Close(); err != nil {
+	if err := join.parentSide.plan.Close(); err != nil {
 		return err
 	}
 
-	return join.subType.Close()
+	return join.childSide.plan.Close()
 }
 
 func (join *invertibleTypeJoin) Spans(spans core.Spans) {
-	join.root.Spans(spans)
+	join.parentSide.plan.Spans(spans)
+}
+
+func (join *invertibleTypeJoin) Source() planNode { return join.parentSide.plan }
+
+type primaryObjectsFetcher struct {
```

Comment:
thought: I have mixed feelings about this being called `fetcher`, but not strong enough to request a change, flagging here in case we want a discussion or if others have stronger thoughts.

Reply:
I have the same feelings. It used to be called "....Joiner" but I extracted join functionality out of this struct, so it is left with fetching-related functionality.
Some suggestions: `primaryObjectsFetchCoordinator`, `primaryObjectsRetriever`
Do you have preferences or suggestions?

Reply:
Out of those two I think I prefer `primaryObjectsRetriever`, although it does feel like a slightly force synonym of `fetcher` :)  Still no strong feelings from me here though.

Reply:
renamed